File: app/src/main/java/com/six/cat/sixcat/SixCatApplication.java
package com.six.cat.sixcat;

import android.app.Application;

import com.akaita.java.rxjava2debug.RxJava2Debug;
import com.squareup.leakcanary.LeakCanary;


/**
 * @author liguoying
 * @date 2017/12/4.
 *
 */

public class SixCatApplication extends Application {
    private static SixCatApplication instance;

    public static SixCatApplication getInstance() {
        SixCatApplication application;
        synchronized (SixCatApplication.class) {
            if (instance == null) {
                synchronized (SixCatApplication.class) {
                    instance = new SixCatApplication();
                }
            }
            application = instance;
        }
        return application;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;
//        if (LeakCanary.isInAnalyzerProcess(this)) {
//            return;
//        }
//        Fabric.with(this, new Crashlytics());

        // Enable RxJava assembly stack collection, to make RxJava crash reports clear and unique
        // Make sure this is called AFTER setting up any Crash reporting mechanism as Crashlytics
        RxJava2Debug.enableRxJava2AssemblyTracking(new String[]{"com.six.cat.sixcat", "com.six.cat.sixcat"});

        LeakCanary.install(this);

    }
}

File: app/src/main/java/com/sixcat/SixCatApplication.kt
package com.sixcat

import android.app.Activity
import android.app.Application
import android.content.Context
import android.os.Bundle
import android.util.Log
import com.akaita.java.rxjava2debug.RxJava2Debug
import com.crashlytics.android.Crashlytics
import com.orhanobut.logger.AndroidLogAdapter
import com.orhanobut.logger.Logger
import com.orhanobut.logger.PrettyFormatStrategy
import com.squareup.leakcanary.LeakCanary
import com.squareup.leakcanary.RefWatcher
import io.fabric.sdk.android.Fabric
import kotlin.properties.Delegates


/**
 * @author liguoying
 * @date 2017/12/4.
 */

class SixCatApplication : Application() {

    private var refWatcher: RefWatcher? = null

    companion object {
        private val TAG = "SixCatApplication"
        var context: Context by Delegates.notNull()
            private set

        fun getRefWatcher(context: Context): RefWatcher? {
            val myApplication = context.applicationContext as SixCatApplication
            return myApplication.refWatcher
        }
    }

    override fun onCreate() {
        super.onCreate()
        context = applicationContext
        Fabric.with(this, Crashlytics())
        refWatcher = setupLeakCanary()
        initLogConfig()
        Fabric.with(this, Crashlytics());

        // Enable RxJava assembly stack collection, to make RxJava crash reports clear and unique
        // Make sure this is called AFTER setting up any Crash reporting mechanism as Crashlytics
        RxJava2Debug.enableRxJava2AssemblyTracking(arrayOf("com.sixcat", "com.sixcat"))

        LeakCanary.install(this)
        registerActivityLifecycleCallbacks(mActivityLifecycleCallbacks)

    }

    private fun initLogConfig() {
        val formatStrategy = PrettyFormatStrategy.newBuilder()
                .showThreadInfo(false)  // 隐藏线程信息 默认：显示
                .methodCount(0)         // 决定打印多少行（每一行代表一个方法）默认：2
                .methodOffset(7)        // (Optional) Hides internal method calls up to offset. Default 5
                .tag("hao_zz")   // (Optional) Global tag for every log. Default PRETTY_LOGGER
                .build()
        Logger.addLogAdapter(object : AndroidLogAdapter(formatStrategy) {
            override fun isLoggable(priority: Int, tag: String?): Boolean {
                return BuildConfig.DEBUG
            }
        })
    }

    private fun setupLeakCanary(): RefWatcher? {
        return if (LeakCanary.isInAnalyzerProcess(this)) {
            RefWatcher.DISABLED
        } else LeakCanary.install(this)
    }

    private val mActivityLifecycleCallbacks = object : Application.ActivityLifecycleCallbacks {
        override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
            Log.d(TAG, "onCreated: " + activity.componentName.className)
        }

        override fun onActivityStarted(activity: Activity) {
            Log.d(TAG, "onStart: " + activity.componentName.className)
        }

        override fun onActivityResumed(activity: Activity) {

        }

        override fun onActivityPaused(activity: Activity) {

        }

        override fun onActivityStopped(activity: Activity) {

        }

        override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {

        }

        override fun onActivityDestroyed(activity: Activity) {
            Log.d(TAG, "onDestroy: " + activity.componentName.className)
        }
    }

}
